src/extensions/compiler/workspace-compiler.ts

```diff
@@ -0,0 +1,173 @@
+  ): Promise<BuildResult[]> {
+    const bitIds = this.getBitIds(componentsIds);
+    const { components } = await this.workspace.consumer.loadComponents(BitIds.fromArray(bitIds));
+    const componentsWithLegacyCompilers: ConsumerComponent[] = [];
+    const componentsAndNewCompilers: ComponentCompiler[] = [];
+    components.forEach((c) => {
+      const environment = this.envs.getEnvFromExtensions(c.extensions);
+      const compilerInstance = environment?.getCompiler();
+      // if there is no componentDir (e.g. author that added files, not dir), then we can't write the dists
+      // inside the component dir.
+      if (compilerInstance && c.componentMap?.getComponentDir()) {
+        const compilerName = compilerInstance.constructor.name || 'compiler';
+        componentsAndNewCompilers.push(new ComponentCompiler(this.workspace, c, compilerInstance, compilerName));
+      } else {
+        componentsWithLegacyCompilers.push(c);
+      }
+    });
+    let newCompilersResultOnWorkspace: BuildResult[] = [];
+    let oldCompilersResult: BuildResult[] = [];
+    if (componentsAndNewCompilers.length) {
+      newCompilersResultOnWorkspace = await BluebirdPromise.mapSeries(
+        componentsAndNewCompilers,
+        (componentAndNewCompilers) => componentAndNewCompilers.compile()
+      );
+    }
+    if (componentsWithLegacyCompilers.length) {
+      oldCompilersResult = await this.compileWithLegacyCompilers(componentsWithLegacyCompilers, options);
+    }
+
+    return [...newCompilersResultOnWorkspace, ...oldCompilersResult];
+  }
+
+  private async compileWithLegacyCompilers(
+    components: ConsumerComponent[],
+    options: LegacyCompilerOptions
+  ): Promise<BuildResult[]> {
+    logger.debugAndAddBreadCrumb('scope.buildMultiple', 'using the legacy build mechanism');
+    const build = async (component: ConsumerComponent) => {
+      if (component.compiler) loader.start(`building component - ${component.id}`);
+      await component.build({
+        scope: this.workspace.consumer.scope,
+        consumer: this.workspace.consumer,
+        ...options,
+      });
+      const buildResults = await component.dists.writeDists(component, this.workspace.consumer, false);
+      if (component.compiler) loader.succeed();
+      return { component: component.id.toString(), buildResults };
+    };
+    const writeLinks = async (component: ConsumerComponent) =>
+      component.dists.writeDistsLinks(component, this.workspace.consumer);
+
+    const buildResults = await BluebirdPromise.mapSeries(components, build);
+    await BluebirdPromise.mapSeries(components, writeLinks);
+
+    return buildResults;
+  }
+
+  private getBitIds(componentsIds: Array<string | BitId>): BitId[] {
+    if (componentsIds.length) {
+      return componentsIds.map((compId) =>
+        compId instanceof BitId ? compId : this.workspace.consumer.getParsedId(compId)
+      );
+    }
+    return this.workspace.consumer.bitMap.getAuthoredAndImportedBitIds();
```

Comment:
add this as api for the consumer (with a proper name), you should almost never use the workspace.consumer. if you use it, it usually means you did a hack and you need to add an API to the workspace